### ISSUE TITLE
Nullable reference types: No unreachable warning for null check on not-nullable variable

### ISSUE THREAD

**Issue by gafter** [dotnet/csharplang#1953]
@bkoelman commented on [Tue Mar 22 2016](https://github.com/dotnet/roslyn/issues/9980)

I have been playing with nullable reference types, which is already working remarkably great! To try it out, I have [patched up](https://github.com/bkoelman/ResharperCodeContractNullability/tree/staticnull) my analyzer to convert Resharper's `(Item)NotNull/CanBeNullAttribute` annotations to nullable reference types. Then ran that on an [actual annotated codebase](https://github.com/bkoelman/DogAgilityCompetitionManagement). 

During that process I discovered some oddities, which I'll create issues for here.

**Version Used**: Built from source, branch NullableReferenceTypes (commit https://github.com/dotnet/roslyn/commit/8553cdac75c2b0901e8ae401aaa23022f0689935)

**Steps to Reproduce**:
The next block of code does not raise any warnings:

``` csharp
void Test()
{
    const string s = "x";

    if (s == null)
    {
        throw new Exception();
    }
}
```

**Expected Behavior**: A warning that indicates the null check will never evaluate to `true`; possibly reuse [CS0472](https://msdn.microsoft.com/en-us/library/ms228621%28v=vs.90%29.aspx). Followed by [CS0162](https://msdn.microsoft.com/en-us/library/c0h4st1x.aspx), which indicates the `throw` line is unreachable.

**Actual Behavior**:
No warnings.

@AlekseyTs 

---

@AlekseyTs commented on [Tue Mar 22 2016](https://github.com/dotnet/roslyn/issues/9980#issuecomment-200057541)

@bkoelman Is the actual behavior accurate? There are no warnings reported at all? 

---

@AlekseyTs commented on [Tue Mar 22 2016](https://github.com/dotnet/roslyn/issues/9980#issuecomment-200058865)

And no errors?

---

@bkoelman commented on [Tue Mar 22 2016](https://github.com/dotnet/roslyn/issues/9980#issuecomment-200062566)

This is what I get (running in VS2015 Update1):

![image](https://cloud.githubusercontent.com/assets/10324372/13969975/35852396-f086-11e5-91dd-d97df9f012d1.png)

---

@bkoelman commented on [Tue Mar 22 2016](https://github.com/dotnet/roslyn/issues/9980#issuecomment-200062840)

Seeing it back, the picture is hard to read. The warning line contains `5`, not `s`.

---

@AlekseyTs commented on [Tue Mar 22 2016](https://github.com/dotnet/roslyn/issues/9980#issuecomment-200064182)

@MadsTorgersen If I remember correctly, we decided against reporting warnings for null checks against values we "know" are not null. The lack of CS0162 is not related to "Nullable Reference Types" feature.

---

@bkoelman commented on [Sun Apr 01 2018](https://github.com/dotnet/roslyn/issues/9980#issuecomment-377797743)

This repros in [Nullable Reference Types Preview 3/6/18](https://github.com/dotnet/csharplang/wiki/Nullable-Reference-Types-Preview).

I'm surprised about that decision, because the description for [CS0472](https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2008/ms228621(v=vs.90)) literally contains:

*The compiler should warn if you use an operator with a constant null value.*

Can I ask what the motivation was against emitting a warning in this case?

**Comment by yaakov-h** [comment]
There's also no warning emitted for comparing structs (which are also guaranteed to be non-null) to `null`:

```csharp
using System;

public class C
{
    public void M(Guid g)
    {
        if (g == null)
        {
        }
        
        if (g != null)
        {
        }
    }
}
```

That can't be added save for warning waves, but for a new feature whose sole purpose is to add warnings, perhaps we should emit a warning for comparing a known-non-null value to null?

**Comment by gafter** [comment]
The nullable analysis feature doesn't "know" when values are not null. It computes when values are *expected* not to be null. For example, an incoming parameter that is declared to be a non-null string still might be null. It is expected and normal for the body of a method to check if the parameter is null and throw an `ArgumentNullException` if it is, even though it was declared not to be null. The purpose of the *nullable reference types* feature is not to warn you about possibly useless code. The purpose is to warn you about code that may fail at runtime due to `null`.

**Comment by CyrusNajmabadi** [comment]
> The nullable analysis feature doesn't "know" when values are not null. It computes when values are expected not to be null

True.  But there are certainly places we know when things truly are not-null, right?  For example, constants.   in @bkoelman 's case:

 
![image](https://user-images.githubusercontent.com/4564579/47394072-6f182980-d6d6-11e8-9806-928fb7324b08.png)

This is a bit surprising to me as well *given* that this is specifically a constant with a non-null value.  That said, while i believe a warning should happen here, i can also see it not specifically being part of NRTs, and more being a general warning wave about superfluous checks.

**Comment by gafter** [comment]
> But there are certainly places we know when things truly are not-null, right?

No, we do not track when variables are "definitely, really, for sure" not null. However, there are no variables in this example, only a declared constant.

@bkoelman 's case is one where you have an if statement with a constant false expression as the condition.  The language specification since C# 1 makes this a case where we treat the body of the controlled statement as unreachable, and we have long (since C# 1?) given a warning on the body

``` c#
    public static void Main()
    {
        const string s = "foo";
        if (s == null)
        {
            Console.WriteLine(s); // warning CS0162: Unreachable code detected
        }
    }
```

The case of testing 5 against null similarly has nothing to do with the nullable reference types feature

``` c#
    public static void Main()
    {
        const string s = "foo";
        if (5 == null) // warning CS0472: The result of the expression is always 'false' since a value of type 'int' is never equal to 'null' of type 'int?'
        {
            Console.WriteLine(s); // warning CS0162: Unreachable code detected
        }
    }
```

In short, we already give warnings for both situations.

**Comment by gafter** [comment]
See [HERE](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8kAmIA1AD4ACADAATkCMA3ALABQr5ATLXQOysDerasNoBmbgDZaAFmoBZCJgB2ACgCUQkYJYjd1AMYB7JbG40o1ALzUARNkOGbzHXuE5qKi5etKArgBt/DRdXbVdXegBOTzVncOoAX01XdxUAVisfAKDkvTD4kSiYuPCkkMTWMqA) for a demonstration that we do indeed give warnings for both situations.
